source/ada/lsp-ada_handlers.adb

```diff
@@ -165,6 +175,54 @@ package body LSP.Ada_Handlers is
    --  Attempt to load the given project file, with the scenario provided.
    --  This unloads all currently loaded project contexts.
 
+   ----------------------
+   -- Contexts_For_URI --
+   ----------------------
+
+   function Contexts_For_URI
+     (Self : access Message_Handler;
+      URI  : LSP.Messages.DocumentUri)
+      return LSP.Ada_Context_Sets.Context_Lists.List
+   is
+      function All_Contexts return LSP.Ada_Context_Sets.Context_Lists.List;
+      --  All known contexts
+
+      function All_Contexts return LSP.Ada_Context_Sets.Context_Lists.List is
+         Result : LSP.Ada_Context_Sets.Context_Lists.List;
+      begin
+         for E in Self.Contexts.Each_Context loop
+            Result.Append (LSP.Ada_Context_Sets.Context_Lists.Element (E));
+         end loop;
+         return Result;
+      end All_Contexts;
+
+      File    : constant Virtual_File := To_File (URI);
+      Result  : LSP.Ada_Context_Sets.Context_Lists.List;
+      Context : Context_Access;
+   begin
+      --  If the file does not exist on disk, assume this is a file
+      --  being created and, as a special convenience in this case,
+      --  assume it could belong to any project.
+      if not File.Is_Regular_File then
+         return All_Contexts;
+      end if;
+
+      --  If the file is a runtime file for the loaded project environment,
+      --  all projects can see it.
+      if Self.Project_Predefined_Sources.Contains (File) then
+         return All_Contexts;
+      end if;
+
+      for E in Self.Contexts.Each_Context loop
```

Comment:
Here will became a call to Self.Contexts.Contexts_For_URI from  lsp-ada_context_sets.ads

Reply:
This doesn't work, unfortunately: the function Contexts_For_URI needs to know the runtime files for the loaded project tree: this is owned by the Ada_Handler, not by the Context_Set.